File: client/Services/Heartbeat.cpp
#include "heartbeat.h"
#include "../network/http_client.h"
#include "../helpers/constants.h"
#include <iostream>
#include <chrono>
#include <thread>

Heartbeat::Heartbeat(const std::string& connectionId, DisconnectCallback onDisconnected)
    : _connectionId(connectionId), _running(true), _onDisconnected(onDisconnected) {
    _thread = std::thread(&Heartbeat::Run, this);
}

Heartbeat::~Heartbeat() {
    Stop();
}

void Heartbeat::Stop() {
    _running = false;
    if (_thread.joinable()) {
        _thread.join();
    }
}

void Heartbeat::Run() {
    int failures = 0;

    while (_running) {
        std::this_thread::sleep_for(std::chrono::milliseconds(HEARTBEAT_INTERVAL));
        if (!_running) break;

        try {
            std::string url = Constants::API_BASE + "/connections/" + _connectionId + "/heartbeat";
            auto response = HttpClient::Post(url, "");

            if (response.statusCode == 200) {
                failures = 0;
            } else if (response.statusCode == 404 || response.statusCode == 0) {
                failures = MAX_FAILURES;
                std::cout << "Heartbeat fatal (" << response.statusCode << ") -> disconnecting immediately" << std::endl;
            } else {
                ++failures;
                std::cout << "Heartbeat failed (" << response.statusCode << ") -> " << failures << "/" << MAX_FAILURES << std::endl;
            }
        }
        catch (const std::exception& ex) {
            ++failures;
            std::cout << "Heartbeat error: " << ex.what() << " -> " << failures << "/" << MAX_FAILURES << std::endl;
        }

        if (failures >= MAX_FAILURES && _running) {
            std::cout << "[disconnected] Heartbeat lost." << std::endl;
            _running = false;
            if (_onDisconnected) {
                std::thread([cb = _onDisconnected]() { cb(); }).detach();
            }
            return;
        }
    }
}

File: client_cs/Services/Heartbeat.cs
using System;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using client.Helpers;

namespace client.Services
{
    public class Heartbeat
    {
        public delegate void DisconnectCallback();

        private readonly HttpClient         _client;
        private readonly string             _connectionId;
        private readonly DisconnectCallback _onDisconnected;
        private readonly Thread             _thread;
        private readonly CancellationTokenSource _cts;
        private volatile bool               _running;

        private const int HEARTBEAT_INTERVAL = 2000;
        private const int MAX_FAILURES       = 3;

        public Heartbeat(string connectionId, DisconnectCallback onDisconnected = null)
        {
            _cts            = new CancellationTokenSource();
            _client         = new HttpClient { Timeout = TimeSpan.FromSeconds(5) };
            _connectionId   = connectionId;
            _onDisconnected = onDisconnected;
            _running        = true;
            _thread         = new Thread(Run) { IsBackground = true };
            _thread.Start();
        }

        public void Stop()
        {
            _running = false;
            _cts.Cancel();
            _client.CancelPendingRequests();
            if (_thread.IsAlive)
                _thread.Join(TimeSpan.FromSeconds(3));
        }

        private void Run()
        {
            int failures = 0;

            while (_running)
            {
                for (int i = 0; i < HEARTBEAT_INTERVAL / 100; i++)
                {
                    if (!_running || _cts.IsCancellationRequested) return;
                    Thread.Sleep(100);
                }

                if (!_running) break;

                try
                {
                    string url      = $"{Constants.API_BASE}/connections/{_connectionId}/heartbeat";
                    var    response = _client.PostAsync(url, null).GetAwaiter().GetResult();
                    int    status   = (int)response.StatusCode;

                    if (status == 200)
                    {
                        failures = 0;
                    }
                    else
                    {
                        failures = MAX_FAILURES;
                        Console.WriteLine($"Heartbeat fatal ({status}) -> disconnecting immediately");
                    }
                }
                catch (Exception ex)
                {
                    if (!_running || _cts.IsCancellationRequested) return;
                    failures = MAX_FAILURES;
                    Console.WriteLine($"Heartbeat fatal: {ex.Message} -> disconnecting immediately");
                }

                if (failures >= MAX_FAILURES && _running)
                {
                    Console.WriteLine("[disconnected] Heartbeat lost.");
                    _running = false;
                    if (_onDisconnected != null)
                    {
                        var cb = _onDisconnected;
                        new Thread(() => cb()) { IsBackground = true }.Start();
                    }
                    return;
                }
            }
        }
    }
}
